flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreITCase.java

```diff
@@ -283,6 +286,54 @@ public void testConcurrentCheckpointOperations() throws Exception {
 		recoveredTestCheckpoint.awaitDiscard();
 	}
 
+	/**
+	 * FLINK-17073 tests that there is no request triggered when there are too many checkpoints
+	 * waiting to clean and that it resumes when the number of waiting checkpoints as gone below
+	 * the threshold.
+	 *
+	 */
+	@Test
+	public void testChekpointingPausesAndResumeWhenTooManyCheckpoints() throws Exception{
+		ManualClock clock = new ManualClock();
+		clock.advanceTime(1, TimeUnit.DAYS);
+		int maxCleaningCheckpoints = 1;
+		CheckpointsCleaner checkpointsCleaner = new CheckpointsCleaner();
+		CheckpointRequestDecider checkpointRequestDecider =  new CheckpointRequestDecider(maxCleaningCheckpoints, unused ->{}, clock, 1, new AtomicInteger(0)::get, checkpointsCleaner::getNumberOfCheckpointsToClean);
+
+		final int maxCheckpointsToRetain = 1;
+		Executors.PausableThreadPoolExecutor executor = Executors.pausableExecutor();
+		ZooKeeperCompletedCheckpointStore checkpointStore = createCompletedCheckpoints(maxCheckpointsToRetain, executor);
+
+		//pause the executor to pause checkpoints cleaning, to allow assertions
+		executor.pause();
+
+		int nbCheckpointsToInject = 3;
+		for (int i = 1; i <= nbCheckpointsToInject; i++) {
+			// add checkpoints to clean
+			TestCompletedCheckpoint completedCheckpoint = new TestCompletedCheckpoint(new JobID(), i,
+				i, Collections.emptyMap(), CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.RETAIN_ON_FAILURE),
+				checkpointsCleaner::cleanCheckpoint);
+			checkpointStore.addCheckpoint(completedCheckpoint);
+		}
+
+		Thread.sleep(100L); // give time to submit checkpoints for cleaning
+
+		int nbCheckpointsSubmittedForCleaningByCheckpointStore = nbCheckpointsToInject - maxCheckpointsToRetain;
+		assertEquals(nbCheckpointsSubmittedForCleaningByCheckpointStore, checkpointsCleaner.getNumberOfCheckpointsToClean());
```

Comment:
This `Thread.sleep` can be fragile. How about replacing it with a loop:
```
while (checkpointsCleaner.getNumberOfCheckpointsToClean() < nbCheckpointsToInject - maxCheckpointsToRetain) {
    Thread.sleep(10);
}
```
?

(ditto: the next `sleep`)

Reply:
agree. I don't like absolute sleeps either but I did not go into so much thinking about that, I first wanted that the ITest behavior is validated before dealing with details.

Reply:
Some thing troubles me though: using the thing we want to test (`checkpointsCleaner.getNumberOfCheckpointsToClean()`) as a condition for a waiting loop seems not a good idea to me as if the tested element fails, we could have an infinite loop in place of a failing ITest. Maybe more trying to synchronise the ITest and checkpointsStore submit somehow (even by raising the visibility of some internals for the purpose of test). 

WDYT ?

Reply:
No, I used a CountDownLatch for synchronising the executor and the test

Reply:
I find `CountDownLatch` approach worse compared to `while` loops because:
1. it's more fragile because ZKStore can submit several `Runnable`s
1. I need to follow when latches are triggered
1. with a new Executor added, it's more complex

For 1, we should probably use `triggerAll` instead of `trigger`. But still it's unclear what should be the counts in latches.
For 3, I think executor could provide methods like `awaitSubmitted(count)` and `awaitAllTriggered`.

But I'd still prefer a loop.

(please unresolve this thread)

Reply:
For 1. you're right, `nbCheckpointsSubmittedForCleaning == 2` so `CountDownLatch(1)` is not correct. +1 on `triggerAll`

For 2. true
For 3. it is just an extend with 2 overrides it does not seem a big deal to me. Moreover, this new class is only in scope test.

Still I don't like looping on the condition that we are actually testing (the count of cleaning CP). IMHO, failing with a timeout if the tested count is incorrect is misleading and makes the build longer so IMHO it is not acceptable. 
I'll try to find a better way to sync the executor and the test rather than the CountDownLatch.

PS: I resolved the discussion by mistake. I just unresolved it.

Reply:
@rkhachatryan I used waiting loops instead of `countDownLatch` to synchronise the test and the executor. And I simply used the executor queue size as the condition for the waiting loop. That way we no more have the condition I did not like and we no more have the countDownLatch and the new executor class you did not like. 
Do you find this solution acceptable ?

Reply:
I think the assumption that each cleanup action is a separate runnable (and that there are no other runnable) can break at some point and is less straightforward than `checkpointsCleaner.getNumberOfCheckpointsToClean() < expected`.

Reply:
Sure the executor can be used for other things that checkpointing in a regular pipeline for the general case. But in the case of this test there is no real pipeline, no `CheckpointCoordinator` just a `ZookeeperCheckpointStore` that gets added some checkpoints. `ZookeeperCheckpointStore` uses the executor only for checkpoints cleaning. In such conditions, I'm not sure that the assumption that there is no other runnables than checkpoint cleanup runnables for the executor does not seem that weak to me.
Do you see a case in which this test environment (only a ZookeeperCPStore created) can generate other runnables for the executor ?

Reply:
Currently not, but I mean possible future changes that can happen in `ZooKeeperCompletedCheckpointStore` or `CheckpointsCleaner`.

Reply:
yes but we are talking about test code not production code. That is not a big deal if a test starts to fail when production code is changed. If that happens, the author will change the test code. Can we merge the PR ?

Reply:
I think forcing the developer to change the tests should be avoided if possible, and here it is possible. 
Also, condition on the public interface rather than internal details makes expectations more clear.

From my side, the PR can be merged after resolving this issue and grooming commits.

Probably some other developers would like to take a brief look - I'll ask about it.

Reply:
I mean, a test reflects the state of the production code at a given time, and at this time a checkpoint cleaning task is submitted as a single runnable. Anyway, if the problem is relying on internals of the executor, I can add await methods in the public API of the executor as you suggested in a previous comment. That way the test author will know that the test expects a certain number of submitted and triggered runnables. I pushed a commit in that sense. Is this ok for you ?

Reply:
The latest commit still makes the assumption about the number of Runnables.
I just dealt with a similar assumption in another test while rebasing my own PR, and would like to avoid it in the future :) 
But let's maybe wait for other developer's feedback. I asked one of my colleagues to take a look at your PR.

Reply:
Yes let's wait for other feedback @tillrohrmann maybe as he knows the ticket very well ?
I also wanted to mention that the aim of this ticket is to count the runnables stacked in the executor (before getting executed) so that they do not lead to OOM. So, it does not choc me to rely on the count of runnables in the test.

Reply:
By the way, we will need another reviewer anyway to merge the PR.

Reply:
I am not entirely sure about which current code snippet we are talking because this one seems to point at some outdated version. Hence I can only give my general thoughts: I think that Roman has a valid point that relying on implementation details such as the number of submitted `Runnables` is more brittle than relying on some public API. I myself have built quite some of these tests and regretted this shortly after when having to refactor these tests. Hence, my recommendation would be to avoid this as much as possible and rather try to rely on the public API of the components under test.

Reply:
Hi @tillrohrmann thanks for your comment ! 

Regarding the code version, we are referring to the last commit (https://github.com/apache/flink/pull/13040/commits/aaf4aca0d5ce666001035f6ae66bca841f87c74d). 

We agreed on using a waiting loop to sync the executor and the test, the only point of discussion is on the condition of this waiting loop: I did not want to use the element that this test tests (the number of cleaning checkpoints) as a condition for a waiting loop. I agree that using internals for this condition is not ideal but do you have any condition to suggest for these waiting loops other than the number of cleaning checkpoints or the number of runnables submitted ?

Reply:
No, I think waiting on the number of checkpoints to clean would be the best option as far as I know. If you are afraid of indefinite waiting you could also use `CommonTestUtils.waitUntilCondition` which allows to specify a timeout for the check.

Reply:
Fair enough, if the condition on internals is too risky and if the condition on nb of checkpoints to clean is the only other solution, I'll do that. Thanks for your opinion guys !